fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -494,18 +494,24 @@ def _setup_backward_hooks(self) -> None:
             if param.grad is not None and param.grad.requires_grad:
                 raise RuntimeError("ShardedDataParallel only works with gradients that don't require grad")
 
-            # Register the hook to the next function in line,
-            # so that the hook is fired when this grad has properly been computed
             p_tmp = param.expand_as(param)
-            assert p_tmp.grad_fn is not None
-            grad_acc = p_tmp.grad_fn.next_functions[0][0]
-            dst_rank = self._trainable_param_to_rank[param]
-
-            reduce_function = self._get_reduce_fn(index, param, dst_rank)
 
-            self._grad_hooks.append(grad_acc.register_hook(reduce_function))
-            self._grad_accs.append(grad_acc)  # keep this hook in scope
-            self._manual_reduce.append(reduce_function)
+            # See https://pytorch.org/docs/stable/tensors.html?highlight=grad_fn
+            # We're interested in the tensors which will be tracked by Autograd
+            # Some tensors can have gradients independent of the inputs (ie. pooling layer for instance),
+            # these do not need to be sync'ed
+            if p_tmp.grad_fn is not None:
```

Comment:
the assert here was too strong, we can just ignore the params whose gradients don't depend on the inputs (see link), by definition they don't need to be synced since all the ranks will have the same. I could not repro an issue here though (even with 1.9 nightly), but several users faced the issue with a recent pytorch, so I'm  guessing that pytorch is getting more aggressive at pruning .grad_fn when not required